[added] Thumbnail component including docs and tests

JordanTheriault opened the pull request:
For Issue #56 Added the thumbnail component, docs, and tests.

Modeled the docs after the [bootstrap doc](http://getbootstrap.com/components/#thumbnails).

I'm new to this repo, so please let me know if you see anything I should change even if it's just best practices. :)

mathieumg commented:
I imagine adding tests would be a good idea! :smile:

JordanTheriault commented:
@mathieumg I've added tests in the commit to test/ThumbnailSpec.js - is there somewhere else I should be adding tests to?

I'm confused regarding the warnings, there seem to be a lot of them!

Looks like there are several linting errors in Thumbnail.js and a few other files.

mathieumg commented:
I skimmed the file names before making my comment, looks like I missed that one... Sorry!

mtscout6 commented on docs/examples/Thumbnail.js:
```diff
@@ -0,0 +1,23 @@
+</Grid>
+);
+
+React.render(thumbnailInstance, mountNode);
```
Can you change your tabs to spaces? I know this is a small thing but it keeps with consistency throughout the project. We have an `.editorconfig` file at the repo root. [Editor Config](http://editorconfig.org/) has plugins for the most popular code editors out there, and will ensure that this style is maintained for you.

mtscout6 commented on docs/src/ComponentsPage.js:
```diff
@@ -448,6 +448,14 @@ const ComponentsPage = React.createClass({
+                <p>Thumbnails are designed to showcase linked images with minimal required markup. You can extend the grid component with thumbnails.</p>
+                  <ReactPlayground codeText={Samples.Thumbnail} />
+                </div>
+
```
Can you put this in the document in the same order as the navigation you used below?

JordanTheriault replied:
Below Grid and above ListGroup or underneath Tabpane and above Tooltip?

mtscout6 replied:
Below Grid and above ListGroup is fine with me.

mtscout6 commented on src/Thumbnail.js:
```diff
@@ -0,0 +1,25 @@
+		let classes = this.getBsClassSet();
+
+		return (
+			<a {...this.props} className={classSet(this.props.className, classes)}>
```
Only using an anchor does not account for other options. The bootstrap docs highlight both `<a>` and `<div>`. See: http://getbootstrap.com/components/#thumbnails

JordanTheriault replied:
What would best practice be for this? Two different tags such as <ThumbnailDiv> and <ThumbnailAnchor> or an attribute?

mtscout6 replied:
I'm thinking that if there is an `href` prop than use an anchor, otherwise use a div. Also, have you considered having `src` and `alt` props and creating the img tag within this component?

So the external api would be something like:

```
<Thumbnail src='...' alt='...' href='...' />

// Which produces:

<a href="..." class="thumbnail">
  <img src="..." alt="...">
</a>
```

OR

```
<Thumbnail src='...' alt='...'>
  <h3>Thumbnail label</h3>
  <p>...</p>
  <p>
    <a href="#" class="btn btn-primary" role="button">Button</a>
    <a href="#" class="btn btn-default" role="button">Button</a>
  </p>
</Thumbnail>

// Which Produces: (Note the caption div)

<div class="thumbnail">
  <img src="..." alt="...">
  <div class="caption">
    <h3>Thumbnail label</h3>
    <p>...</p>
    <p><a href="#" class="btn btn-primary" role="button">Button</a> <a href="#" class="btn btn-default" role="button">Button</a></p>
  </div>
</div>
```

JordanTheriault replied:
Sounds like a great way of differentiating, I will get to work on updating this - thank you!

mtscout6 commented on src/Thumbnail.js:
```diff
@@ -0,0 +1,37 @@
+        <div {...this.props} className={classSet(this.props.className, classes)}>
+          <img src={this.props.src} alt={this.props.alt} />
+          <div className="caption">
+            {this.props.children}
```
If there are no children do we even need this div?

mtscout6 replied:
Looks like there's too much padding with no children

![](http://cl.ly/image/3R2D1S331M1H/Image%202015-04-16%20at%204.50.02%20PM.png)

JordanTheriault replied:
I will change this to ensure that if there are not children, the caption div will not be created.

mtscout6 commented on src/Thumbnail.js:
```diff
@@ -0,0 +1,37 @@
+
+    if(this.props.href) {
+      return (
+        <a {...this.props} href={this.props.href} className={classSet(this.props.className, classes)}>
```
If `this.props.href` is defined and there are children should we warn that the children will not be rendered or render them within this anchor? I'm not sure how Bootstrap does this, or what they prescribe, thoughts?

dozoisch replied:
I think we should warn, always better to let people know

JordanTheriault replied:
By warn, do you mean in the docs, to console, or some other method?

mtscout6 replied:
If we just warn, that's not very enabling.

I see two options:

1. Render the anchor tag around the image only.
2. Render the anchor tag around the image and it's children.

Option 1 is probably the safest since it's entirely possible to pass a button or anchor as a child. The question is though, what is the prescribed Bootstrap solution or is this one up to interpretation by us?

dozoisch commented:
@JordanTheriault If you can answer the lastest comments, and rebase and we will be able to merge!

JordanTheriault commented:
@dozoisch Back from vacation, I will respond then rebase once it's determined its ready to go.

dozoisch commented:
@JordanTheriault Awesome!

JordanTheriault commented:
I've rebased and fixed the issue @mtscout6 mentioned above with the spacing when no children are present. I'm receiving warnings regarding self-closing components.

```
test/ThumbnailSpec.js
  23:6  warning  Empty components are self-closing  react/self-closing-comp
  32:6  warning  Empty components are self-closing  react/self-closing-comp
  40:6  warning  Empty components are self-closing  react/self-closing-comp
```

Can anyone advise on how I can change the ThumbnailSpec to fix these warnings?

mtscout6 commented:
@JordanTheriault Change those to `<Thumbnail src="#" alt="test" />`

JordanTheriault commented:
Although this is failing Travis, it passes lint/test/etc locally. Any advice?

mtscout6 commented:
I tried to send you a shout out on [gitter](https://gitter.im/react-bootstrap/react-bootstrap?at=555514cdb2001f92677cb9e8), this is because of babel/babel-eslint#96.

mtscout6 commented:
It passes locally for me as well. Can you squash the commits?

mtscout6 commented:
Sorry for the turn around but can you amend your commit message to follow the commit message guidelines as prescribed in the [Contributing Guide](https://github.com/react-bootstrap/react-bootstrap/blob/master/CONTRIBUTING.md#commit-subjects-for-public-api-changes)? This will ensure that this change is documented in the changelog on the next release.
